frontend/utils/crypto.js

```diff
@@ -1,51 +1,324 @@
 'use strict'
 
+import nacl from 'tweetnacl'
+
+import { blake32Hash, bytesToB64, b64ToBuf, strToBuf } from '~/shared/functions.js'
+
 import scrypt from 'scrypt-async'
 
-export class Key {
-  constructor (privKey, pubKey, salt) {
-    this.privKey = privKey
-    this.pubKey = pubKey // optional
-    this.salt = salt // optional
+export type Key = {
+  type: string;
+  secretKey?: any;
+  publicKey?: any;
+}
+
+export const keygen = (type: string): Key => {
+  if (type === 'edwards25519sha512batch') {
+    const key = nacl.sign.keyPair()
+
+    const res: Key = {
+      type: type,
+      publicKey: key.publicKey
+    }
+
+    Object.defineProperty(res, 'secretKey', { value: key.secretKey })
```

Comment:
Why `Object.defineProperty` here instead of using simpler / more common syntax?

Reply:
To make it 'hidden' and thus harder to accidentally leak.

```
> var x = {a: 'b'}; Object.defineProperty(x, 'foo', {value: 'bar'}); JSON.stringify(x)
< '{"a":"b"}'
```

As opposed to:

```
> var x = {a: 'b', 'foo': 'bar'}; JSON.stringify(x)
< '{"a":"b","foo":"bar"}'
```

Reply:
Ah interesting, can you add a comment to that effect?